axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/CartLineSupplyChainServiceImpl.java

```diff
@@ -0,0 +1,81 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.axelor.apps.supplychain.service;
+
+import com.axelor.apps.base.service.app.AppBaseService;
+import com.axelor.apps.sale.db.Cart;
+import com.axelor.apps.sale.db.CartLine;
+import com.axelor.apps.sale.db.repo.SaleOrderLineRepository;
+import com.axelor.apps.stock.db.StockLocationLine;
+import com.axelor.apps.stock.service.StockLocationLineFetchService;
+import com.axelor.i18n.I18n;
+import com.google.inject.Inject;
+import java.math.BigDecimal;
+
+public class CartLineSupplyChainServiceImpl implements CartLineSupplyChainService {
+
+  protected StockLocationLineFetchService stockLocationLineFetchService;
+  protected AppBaseService appBaseService;
+
+  @Inject
+  public CartLineSupplyChainServiceImpl(
+      StockLocationLineFetchService stockLocationLineFetchService, AppBaseService appBaseService) {
+    this.stockLocationLineFetchService = stockLocationLineFetchService;
+    this.appBaseService = appBaseService;
+  }
+
+  @Override
+  public CartLine setAvailableStatus(Cart cart, CartLine cartLine) {
+    BigDecimal availableStock = getAvailableStock(cart, cartLine);
+    BigDecimal allocatedStock = getAllocatedStock(cart, cartLine);
+    BigDecimal availableQty = availableStock.add(allocatedStock);
+    BigDecimal realQty = cartLine.getQty();
+
+    if (availableQty.compareTo(realQty) >= 0) {
+      cartLine.setAvailableStatus(I18n.get("Available"));
+      cartLine.setAvailableStatusSelect(SaleOrderLineRepository.STATUS_AVAILABLE);
+    } else if (availableQty.compareTo(realQty) < 0) {
+      BigDecimal missingQty =
+          availableQty.subtract(realQty).setScale(appBaseService.getNbDecimalDigitForQty());
+      cartLine.setAvailableStatus(I18n.get("Missing") + " (" + missingQty + ")");
+      cartLine.setAvailableStatusSelect(SaleOrderLineRepository.STATUS_MISSING);
+    }
+    return cartLine;
+  }
+
+  protected BigDecimal getAvailableStock(Cart cart, CartLine cartLine) {
+    StockLocationLine stockLocationLine =
+        stockLocationLineFetchService.getStockLocationLine(
+            cart.getStockLocation(), cartLine.getProduct());
+    if (stockLocationLine == null) {
+      return BigDecimal.ZERO;
+    }
+    return stockLocationLine.getCurrentQty().subtract(stockLocationLine.getReservedQty());
+  }
+
+  protected BigDecimal getAllocatedStock(Cart cart, CartLine cartLine) {
+    StockLocationLine stockLocationLine =
+        stockLocationLineFetchService.getStockLocationLine(
+            cart.getStockLocation(), cartLine.getProduct());
+    if (stockLocationLine == null) {
+      return BigDecimal.ZERO;
+    }
+    return stockLocationLine.getReservedQty();
+  }
```

Comment:
you can use `stockLocationLineFetchService.getAvailableQty(StockLocation stockLocation, Product product)`, it will manage stock reservation.